urldee901fae6_DocX_personal_timetabling_tgz-pJ8-net_personaltt_main_BenchmarkJ8/src/net/personaltt/timedomain/ActionStackDomain.java

```diff
@@ -0,0 +1,246 @@
+        Interval boundary = null;
+        
+        // go from ground and build bounding
+        // addition of bounded set adds to boundary
+        // mask of bounded set intersect boundary
+        // removing of bounded set can make boundary smaller, but it will still be boundary
+        //  so do nothing
+        //
+        
+        // first make sure that this is bounded
+        if (this.isBounded() == false) {
+            return null;
+        }
+        
+        for (ActionBase actionBase : actionStack) {
+            boundary = actionBase.applyOnBoundary(boundary);
+        }
+        
+        return boundary;
+    }    
+
+    @Override
+    public boolean intersects(IIntervalsTimeDomain other) {
+        return IntervalsTimeDomainUtils.genericIntersects(this, other);
+    }
+    
+    
+    /* actions implementation */
+    
+    abstract class ActionBase {
+        IIntervalsTimeDomain domain;
+
+        public ActionBase(IIntervalsTimeDomain domain) {
+            this.domain = domain;
+        }
+        
+        // aplies action on given interval set in the given range
+        abstract void applyOn(IntervalsSet s, Interval range);
+        
+        // return true if result of applyon of other on bounded will be bounded
+        abstract boolean boundedAppliedOn(boolean groundBounded);
+        
+        // apply action operand boundary to given boundary using 
+        // action operation and return boundary of result
+        // Return null if action operand is not bounded
+        abstract Interval applyOnBoundary(Interval boundary);
+    }
+    
+    /**
+     * Add action represents union operation of its domain to another
+     */
+    class AddAction extends ActionBase {
+
+        public AddAction(IIntervalsTimeDomain domain) {
+            super(domain);
+        }
+        
+        @Override
+        void applyOn(IntervalsSet s, Interval range) {
+            IntervalsSet i = domain.getIntervalsIn(range);
+            s.unionWith(i);
+        }
+
+        @Override
+        boolean boundedAppliedOn(boolean otherBounded) {
+            boolean thisBounded = domain.isBounded();
+            return thisBounded && otherBounded;
+        }
+
+        @Override
+        Interval applyOnBoundary(Interval boundary) {
+            if (!domain.isBounded()) {
+                return null;
+            }
+                
+            // extends boundary by this domain boundary
+            Interval thisBoundary = this.domain.getBoundingInterval();
+            
+            if (boundary == null) {
+                return thisBoundary;
+            }
+            else {
+                return new Interval(
+                    thisBoundary.getStart().isBefore(boundary.getStart()) ? thisBoundary.getStart() : boundary.getStart(),
+                    thisBoundary.getEnd().isAfter(boundary.getEnd()) ? thisBoundary.getEnd() : boundary.getEnd()
+                    );
+            }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 156:164, 232:241.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.